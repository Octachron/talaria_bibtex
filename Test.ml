open Bibtex
let p=Printf.printf
open BibtexP
open MicroP
let f=open_in "test.bib"


let rec scanp lexb= match MicroL.pages lexb with
	| MicroP.EOF -> p "EOF \n"
	| MicroP.NUM(n) -> p "Num(%d)" n; scanp lexb
	| MicroP.MINUS -> p "<->"; scanp lexb
	| MicroP.WORD(s)-> p "WORD(%s)\n" s; scanp lexb
	| MicroP.COMMA -> p"<,>"
	| MicroP.AND -> p"<and>"
	| MicroP.SEP -> p "</>"; scanp lexb




let bib = Bibparse.parse @@ Lexing.from_channel f
  

let rec scan lexbuf= match BibtexL.main lexbuf with
	| EOF -> p "EOF \n"
	| TEXT(s) -> p "text: \"%s\" \n" s ; scan lexbuf
	| LCURL -> p "<" ; scan lexbuf
	| RCURL -> p ">"; scan lexbuf
	| COMMA -> p ":,:"; scan lexbuf
	| KIND s->p "Type \"%s\"" s; scan lexbuf
	| EQUAL -> p "="; scan lexbuf

let mayr (type p) entry (module M:Orec.Repr.Sig with type p=p) f =  try f @@ M.get entry with
| Not_found -> ()  

let ( *? ) = mayr



let mayp (type p) (module M:Orec.Repr.Sig with type p=p) fmt entry= 
entry *? (module M ) @@ p fmt 

let decorated op en pr e = print_string op; pr e; print_string en

let decoratedN name= decorated (Printf.sprintf "\t %s: [" name) "]\n"

let plist sep pr l=
	let rec inner l = match l with
	| [] -> ()
	| [a] -> pr a 
	| a::q -> pr a; p sep; inner q   in
	inner l

let plistn name pr = decoratedN name @@ plist ", "  pr

let () = Database.iter (fun id entry ->
	p "Entry : %s \n " id;
	entry *?  (module Title) @@ decoratedN "Title" print_string;    
	entry *? (module Authors) @@ plistn "Authors" (fun {firstname;lastname} -> p "%s %s" firstname lastname);
	entry *? (module Year) @@ decoratedN "Year" print_int;
	entry *? (module Journal) @@  decoratedN "Journal" print_string;
	entry *? (module Volume) @@  decoratedN "Volume" print_int;
	entry *? (module Number) @@  decoratedN "Number" print_int;
	entry *? (module Pages) @@ (function Loc k -> p "\t Pages: [%d] \n" k | Interv(k,l) -> p "\t Pages: [%d-%d] \n" k l);
	entry *? (module Arxiv) @@  decoratedN "Arxiv" print_string;
	entry *? (module Doi) @@  plistn "Doi" print_string;
	entry *? (module Tags)  @@  plistn "Tags"  (p "%s"); 
	entry *? (module Abstract) @@  decoratedN "Abstract" print_string;
	
) bib
