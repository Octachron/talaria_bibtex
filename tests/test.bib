@article{ Nothingness,
  title = {On empty articles},
  author = {Nobody, U. and Nonymous, A.},
  year ={1995},
  pages ={0-0}
}
