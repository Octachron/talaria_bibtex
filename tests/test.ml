open Bibtex
let fp =Format.fprintf
let p = Format.printf
let string = Format.pp_print_string
let int = Format.pp_print_int

module Scan_test = struct
  [@@@warning "-unused-value-declaration"]

  let rec scanp ppf lexb= match MicroL.pages lexb with
	| MicroP.EOF -> fp ppf "EOF \n"
	| MicroP.NUM(n) -> fp ppf "Num(%d)" n; scanp ppf lexb
	| MicroP.MINUS -> fp ppf "<->"; scanp ppf lexb
	| MicroP.WORD(s)-> fp ppf "WORD(%s)\n" s; scanp ppf lexb
	| MicroP.COMMA -> fp ppf "<,>"
	| MicroP.AND -> fp ppf "<and>"
	| MicroP.SEP -> fp ppf "</>"; scanp ppf lexb



  let rec scan ppf lexbuf= match Lexer.main lexbuf with
	| EOF -> fp ppf "EOF \n"
	| TEXT(s) -> fp ppf "text: \"%s\" \n" s ; scan ppf lexbuf
	| LCURL -> fp ppf "<" ; scan ppf lexbuf
	| RCURL -> fp ppf ">"; scan ppf lexbuf
	| COMMA -> fp ppf ":,:"; scan ppf lexbuf
	| KIND s-> fp ppf "Type \"%s\"" s; scan ppf lexbuf
	| EQUAL -> fp ppf "="; scan ppf lexbuf
end


let mayr entry field f = let open Fields in  match entry.%{field.f} with
| None -> ()
| Some v -> f Format.std_formatter v

let ( *? ) = mayr


let decorated op en pr ppf e = Format.fprintf ppf "%s%a%s" op pr e en

let decoratedN name= decorated (Format.asprintf "\t %s: [" name) "]\n"

let plist sep = Format.pp_print_list ~pp_sep:(fun ppf () -> Format.fprintf ppf sep)

let plistn name pr = decoratedN name @@ plist ", "  pr

let setp name pr ppf s= s |> Fields.StrSet.elements |>  plistn name pr ppf

let pp_raw ppf entry = match Fields.(entry.%{raw}) with
  | None -> ()
  | Some dtb ->
    if Database.is_empty dtb then ()
    else
      let pp_binding ppf (key,x) = fp ppf "%s={%s}" key x in
      Format.fprintf ppf "raw={%a}" (plist ",@ " pp_binding) (Database.bindings dtb)

let () =
  let f=open_in "test.bib" in
  let bib = parse @@ Lexing.from_channel f in
  let open Fields in
  Database.iter (fun id entry ->
      p "Entry : %s \n " id;
      entry *? title @@ decoratedN "Title" string;
      entry *? authors
      @@ plistn "Authors"
        (fun ppf {firstname;lastname} -> fp ppf "%s %s" firstname lastname);
      entry *? year @@ decoratedN "Year" int;
      entry *? journal @@  decoratedN "Journal" string;
      entry *? volume @@  decoratedN "Volume" int;
      entry *? number @@  decoratedN "Number" int;
      entry *? pages @@
      (fun ppf -> function Loc k -> fp ppf "\t Pages: [%d] \n" k
              | Interv(k,l) -> fp ppf "\t Pages: [%d-%d] \n" k l
      );
      entry *? arxiv @@  decoratedN "Arxiv" string;
      entry *? doi @@  plistn "Doi" string;
      entry *? tags  @@  setp "Tags" string;
      entry *? abstract @@  decoratedN "Abstract" string;
      pp_raw Format.std_formatter entry
    ) bib
