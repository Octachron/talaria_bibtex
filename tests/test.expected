Entry : Nothingness 
 	 Title: [On empty articles]
	 Authors: [U. Nobody, A. Nonymous]
	 Year: [1995]
	 Pages: [0-0] 
